substrate/frame/staking/src/ledger.rs

```diff
@@ -203,9 +218,16 @@ impl<T: Config> StakingLedger<T> {
 
 	/// Clears all data related to a staking ledger and its bond in both [`Ledger`] and [`Bonded`]
 	/// storage items and updates the stash staking lock.
+	///
+	/// Note: we assume that [`T::EventListeners::on_nominator_remove`] and/or
+	/// [`T::EventListeners::on_validator_remove`] have been called and the stash is idle, prior to
+	/// call this method. This can be achieved by calling `[crate::Pallet::<T>::kill_stash]` prior
+	/// to this call.
 	pub(crate) fn kill(stash: &T::AccountId) -> Result<(), Error<T>> {
 		let controller = <Bonded<T>>::get(stash).ok_or(Error::<T>::NotStash)?;
 
+		ensure!(crate::Pallet::<T>::status(stash) == Ok(StakerStatus::Idle), Error::<T>::BadState);
```

Comment:
I will check defensive here instead.

Reply:
While I understand this is a real concern, I'm not convinced that this check should happen here. The ledger itself has no notion of `StakerStatus`. It's actually the only place in `ledger.rs` where you query the status. I'm not saying we shouldn't care about overall state validity, I'm saying this particular piece of code shouldn't care about the notion of `StakerStatus` because it's not responsible for it. Callers owning and mutating the `StakerStatus` state should ensure the state is valid.

I still think it makes a lot more sense to do the status check (`ensure!(Self::status(stash) == Ok(StakerStatus::Idle), Error::<T>::BadState);`), or a `debug_assert`, in `kill_stash` before calling `StakingLedger::kill`.